app/src/main/java/edu/uw/covidsafe/importlocations/ImportGoogleLocations.java

```diff
@@ -0,0 +1,165 @@
+        final Context context = getApplicationContext();
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        // Takeout page: https://takeout.google.com/settings/takeout/custom/location_history
+        // KML page: "https://www.google.com/maps/timeline/kml?authuser=0&pb=!1m8!1m3!1i" + year + "!2i" + month + "!3i" + mday +
+        //        "!2m3!1i" + year + "!2i" + month + "!3i" + mday;
+        // Ex: "https://www.google.com/maps/timeline/kml?authuser=0&pb=!1m8!1m3!1i" + 2020 + "!2i" + 01 + "!3i" + 01 +
+        //                "!2m3!1i" + 2020 + "!2i" + 04 + "!3i" + 01
+        final String signin_url = "https://accounts.google.com/signin/v2/identifier?flowName=GlifWebSignIn&flowEntry=ServiceLogin";
+        final String download_url = "https://www.google.com/maps/timeline/kml?authuser=0&pb=!1m8!1m3!1i" + 2020 + "!2i" + 01 + "!3i" + 01 +
+                "!2m3!1i" + 2020 + "!2i" + 01 + "!3i" + 04;
+        final Map<String, String> headers = new HashMap<>();
+        final WebView myWebView = (WebView) findViewById(R.id.webview);
+
+        headers.put("content-type", "application/vnd.google-earth.kml+xml");
+        WebSettings webSettings = myWebView.getSettings();
+        webSettings.setJavaScriptEnabled(true);
+
+        myWebView.setWebViewClient(new MyWebViewClient() {
+            @Override
+            public void onPageFinished(WebView view, String url) {
+                super.onPageFinished(myWebView, url);
+                if (url.contains("https://myaccount.google.com/")) {
+                    checkStoragePermission();
+                    myWebView.loadUrl(download_url, headers);
+                    myWebView.setVisibility(View.GONE);
+                    myWebView.clearFormData();
+                    myWebView.clearCache(true);
+                    try {
+                        Thread.sleep(5000);
+                        System.out.println("Awake");
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+
+
+                    LocationDataXMLParser parser = new LocationDataXMLParser();
+                    List<List<String>> output = parser.getLinks(context.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS));
+                    for(List<String> section : output) {
+                        for(String text : section) {
+                            System.out.println("Entry:" + text);
+                        }
+                    }
+                }
+            }
+        });
+
+        myWebView.setDownloadListener(new DownloadListener() {
+            @Override
+            public void onDownloadStart(String url, String userAgent,
+                                        String contentDisposition, String mimetype,
+                                        long contentLength) {
+                DownloadManager.Request request = new DownloadManager.Request(

```

Comment:
This is a good use case for intent service . let's discuss offline , will help you with this :)